mmdet3d/models/voxel_encoders/voxel_encoder.py

```diff
@@ -486,3 +486,124 @@ def fusion_with_mask(self, features, mask, voxel_feats, coors, img_feats,
+        self.point_cloud_range = point_cloud_range
+
+        feat_channels = [self.in_channels] + list(feat_channels)
+        vfe_layers = []
+        norm_name, norm_layer = build_norm_layer(norm_cfg, self.in_channels)
+        vfe_layers.append(norm_layer)
+        for i in range(len(feat_channels) - 1):
+            in_filters = feat_channels[i]
+            out_filters = feat_channels[i + 1]
+            norm_name, norm_layer = build_norm_layer(norm_cfg, out_filters)
+            if i == len(feat_channels) - 1:
+                vfe_layers.append(nn.Linear(in_filters, out_filters))
+            else:
+                vfe_layers.append(
+                    nn.Sequential(
+                        nn.Linear(in_filters, out_filters), norm_layer,
+                        nn.ReLU(inplace=True)))
+        self.vfe_layers = nn.ModuleList(vfe_layers)
+        self.num_vfe = len(vfe_layers)
+        self.vfe_scatter = DynamicScatter(voxel_size, point_cloud_range,
+                                          (mode != 'max'))
+        self.compression_layers = None
+        if feat_compression is not None:
+            self.compression_layers = nn.Linear(feat_channels[-1],
+                                                feat_compression)
+
+    def forward(self, features: torch.Tensor, coors: torch.Tensor, *args,
+                **kwargs):
+        """Forward functions.
+
+        Args:
+            features (torch.Tensor): Features of voxels, shape is NxC.
+            coors (torch.Tensor): Coordinates of voxels, shape is  Nx(1+NDim).
+
+        Returns:
+            tuple: If `return_point_feats` is False, returns voxel features and
+                its coordinates. If `return_point_feats` is True, returns
+                feature of each points inside voxels additionally.
+        """
+        features_ls = [features]
+
+        # Find distance of x, y, and z from voxel center
+        if self._with_voxel_center:
+            f_center = features.new_zeros(size=(features.size(0), 3))
+            f_center[:, 0] = features[:, 0] - (
+                coors[:, 3].type_as(features) * self.vx + self.x_offset)
+            f_center[:, 1] = features[:, 1] - (
+                coors[:, 2].type_as(features) * self.vy + self.y_offset)
+            f_center[:, 2] = features[:, 2] - (
+                coors[:, 1].type_as(features) * self.vz + self.z_offset)
+            features_ls.append(f_center)
+
+        # Combine together feature decorations
+        point_feats = torch.cat(features_ls[::-1], dim=-1)
+        for i, vfe in enumerate(self.vfe_layers):
+            point_feats = vfe(point_feats)
+            if i == self.num_vfe - 1:
+                voxel_feats, voxel_coors = self.vfe_scatter(point_feats, coors)
+        if self.compression_layers is not None:
+            voxel_feats = self.compression_layers(voxel_feats)
+
+        if self.return_point_feats:
+            return voxel_feats, voxel_coors, point_feats
```

Comment:
For returned point_feats, it seems to be a list of Tensor output from each `vfe_layers`.